### PR TITLE
SampleData: Fix support for custom downloader

### PR DESCRIPTION
**Pull request by lopezmt**
This is the first step to make it easier for the user to get started with SlicerSALT as suggested in https://github.com/Kitware/SlicerSALT/issues/119

### CODE REVIEW

File: Modules/Scripted/SampleData/SampleData.py [hunk 1]
```diff
@@ -315,6 +315,14 @@ def logMessage(self, message, logLevel=logging.INFO):
     logging.log(logLevel, message)
     slicer.app.processEvents(qt.QEventLoop.ExcludeUserInputEvents)
 
+  def setCategoryVisible(self,categoryName, visible):
+    for i in range(self.layout.count()):
+      widget = self.layout.itemAt(i).widget()
+      if hasattr(widget, "name") and widget.name == '%sCollapsibleGroupBox' % categoryName:
+        if visible:
```
Comment (jcfr) [inline Modules/Scripted/SampleData/SampleData.py]:
```
widget.setVisible(visible)
```

File: Modules/Scripted/SampleData/SampleData.py [hunk 2]
```diff
@@ -315,6 +315,14 @@ def logMessage(self, message, logLevel=logging.INFO):
     logging.log(logLevel, message)
     slicer.app.processEvents(qt.QEventLoop.ExcludeUserInputEvents)
 
+  def setCategoryVisible(self,categoryName, visible):
```
Comment (jcfr) [inline Modules/Scripted/SampleData/SampleData.py]: `(self,categoryName, visible):` -> `(self, categoryName, visible):`

Comment (jcfr) [inline Modules/Scripted/SampleData/SampleData.py]: add docstring

```
"""Update visibility of a SampleData category given its name.
"""
```

Comment (jcfr) [inline Modules/Scripted/SampleData/SampleData.py]: I suggest you introduce a

```
def isRegisteredDataSource(category):
  return category in slicer.modules.sampleDataSources.keys()
```

and then use it in the function

File: Modules/Scripted/SampleData/SampleData.py [hunk 3]
```diff
@@ -315,14 +315,21 @@ def logMessage(self, message, logLevel=logging.INFO):
...
-          widget.hide()
+  def setCategoryVisible(self, categoryName, visible):
+    """Update visibility of a SampleData category given its name.
+    """
+    if self.isRegisteredDataSource(categoryName):
+      for i in range(self.layout.count()):
+        widget = self.layout.itemAt(i).widget()
+        if hasattr(widget, "name") and widget.name == '%sCollapsibleGroupBox' % categoryName:
+          widget.setVisible(visible)
+          return
```
Comment (jcfr) [inline Modules/Scripted/SampleData/SampleData.py]: I suggest to simplify even further  to something like this:

```python
  def setCategoryVisible(self, categoryName, visible):
    """Update visibility of a SampleData category given its name.
    """
    if not self.isRegisteredDataSource(categoryName):
      return
    slicer.util.findChild(self.parent, '%sCollapsibleGroupBox' % categoryName).setVisible(visible)
```

Comment (lopezmt) [reply Modules/Scripted/SampleData/SampleData.py]: This is a really useful function! is there any documentation somewhere that lists all the function in slicer.util?

Comment (lassoan) [reply Modules/Scripted/SampleData/SampleData.py]: See https://slicer.readthedocs.io/en/latest/developer_guide/slicer.html

**Comment by lopezmt** [comment]
This is even shorter, is it still better to split the test and the action or is it fine in one block like that?

```python 
def setCategoryVisible(self, categoryName, visible):
    """Update visibility of a SampleData category given its name.
    """
    if self.isRegisteredDataSource(categoryName):
      slicer.util.findChild(self.parent, '%sCollapsibleGroupBox' % categoryName).setVisible(visible)
```

File: Modules/Scripted/SampleData/SampleData.py (hunk 3 above)
Comment (lopezmt) [reply Modules/Scripted/SampleData/SampleData.py]: This is even shorter, is it still better to split the test and the action or is it fine in one block like that?

```python 
def setCategoryVisible(self, categoryName, visible):
    """Update visibility of a SampleData category given its name.
    """
    if self.isRegisteredDataSource(categoryName):
      slicer.util.findChild(self.parent, '%sCollapsibleGroupBox' % categoryName).setVisible(visible)
```

Comment (jcfr) [reply Modules/Scripted/SampleData/SampleData.py]: It is usually more readable to split things.
```
# check, early exit if needed
# then action
```

**Comment by jcfr** [comment]
* Consider squashing all three commit together and then force-pushing the topic.
* Prefer imperative when writing commit message. See https://www.slicer.org/wiki/Documentation/Nightly/Developers/Style_Guide#Commits
* use `category` instead of `categoryName` (to be consistent with remaining of the code, I incorrectly suggested this earlier)
* add `isCategoryVisible(category)`
* add `test_widget_categoryVisiblity`

**Comment by lopezmt** [comment]
I have never created a test function to test a function, can you tell me what needs to be check to consider that a function is tested?

**Comment by jcfr** [comment]
Thanks for adding tests :+1:  

I suggest to use `isRegisteredCategory` instead of `isRegisteredDataCategory` 

> customDownloader works and remains working?

To follow up on what we discussed earlier today, consider changing this function:

```
  def __init__(self, logMessage=None):
    if logMessage:
      self.logMessage = logMessage
    self.builtInCategoryName = 'BuiltIn'
    self.developmentCategoryName = 'Development'
    self.registerBuiltInSampleDataSources()
    self.registerDevelopmentSampleDataSources()
```

into 

```
  def __init__(self, logMessage=None):
    if logMessage:
      self.logMessage = logMessage
    self.builtInCategoryName = 'BuiltIn'
    self.developmentCategoryName = 'Development'
    self.registerBuiltInSampleDataSources()
    self.registerDevelopmentSampleDataSources()
    if slicer.app.testingEnabled():
      self.registerTestingDataSources()
```

Then implement `registerTestingDataSources`  to register data source with custom downloader and add a a test function `test_dataSourceWithCustomDownloader` that would check it works as expected. The custom downloader could simply raise an RuntimeException and the you would use https://docs.python.org/3.6/library/unittest.html#unittest.TestCase.assertRaises

**Comment by jcfr** [comment]
The support for show/hide of category has been integrated in [r28314](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=28314)

The remaining of the topic has been rebased and updated against the latest Slicer.

@lopezmt For other development, considering updating your Slicer checkout.

File: Modules/Scripted/SampleData/SampleData.py [hunk 4]
```diff
@@ -483,6 +490,16 @@ def registerDevelopmentSampleDataSources(self):
       checksums=['SHA256:c0743772587e2dd4c97d4e894f5486f7a9a202049c8575e032114c0a5c935c3b', 'SHA256:3243b62bde36b1db1cdbfe204785bd4bc1fbb772558d5f8cac964cda8385d470']
       )
 
+  def registerTestingDataSources(self):
+    """Fills in the sample data sources displayed only if developer mode is enabled."""
+
+    self.registerCustomSampleDataSource(
+      category=self.testingCategoryName, sampleName='customDownloader',
```
Comment (jcfr) [inline Modules/Scripted/SampleData/SampleData.py]: Look like `self.testingCategoryName` is invalid

You could simply use a string like `"Testing"`

File: Modules/Scripted/SampleData/SampleData.py [hunk 5]
```diff
@@ -965,3 +983,13 @@ def test_categoryVisibility(self):
     self.assertFalse(widget.isCategoryVisible('BuiltIn'))
     widget.setCategoryVisible('BuiltIn', True)
     self.assertTrue(widget.isCategoryVisible('BuiltIn'))
+
+  def test_customDownloader(self):
```
Comment (jcfr) [inline Modules/Scripted/SampleData/SampleData.py]:
```
slicer.util.selectModule("SampleData")
```

File: Modules/Scripted/SampleData/SampleData.py [hunk 6]
```diff
@@ -965,3 +983,13 @@ def test_categoryVisibility(self):
...
+    logic = SampleDataLogic()
+    widget = slicer.modules.sampledata.widgetRepresentation().self()
```
Comment (jcfr) [inline Modules/Scripted/SampleData/SampleData.py]: -> `widget = slicer.modules.SampleDataWidget`

File: Modules/Scripted/SampleData/SampleData.py [hunk 7]
```diff
@@ -415,13 +415,20 @@ def registerCustomSampleDataSource(category='Custom',
       customDownloader=customDownloader,
       ))
 
+    
+  @staticmethod
+  def CustomDownloadTest(source):
+    raise Exception('CustomDownloadTesting')
```
Comment (jcfr) [inline Modules/Scripted/SampleData/SampleData.py]: Use  `RuntimeError`, `Exception` is too general and the test could instead catch an unrelated exception

Comment (jcfr) [reply Modules/Scripted/SampleData/SampleData.py]: `CustomDownloadTesting` -> `CustomDownloadTesting: This exception confirms that setting a custom download works`

File: Modules/Scripted/SampleData/SampleData.py [hunk 8]
```diff
@@ -415,13 +415,20 @@ def registerCustomSampleDataSource(category='Custom',
       customDownloader=customDownloader,
       ))
 
+    
+  @staticmethod
+  def CustomDownloadTest(source):
```
Comment (jcfr) [inline Modules/Scripted/SampleData/SampleData.py]: Compare source parameter is the registered source and raise an `AssertError` if not the expected one.

File: Modules/Scripted/SampleData/SampleData.py [hunk 9]
```diff
@@ -483,6 +490,16 @@ def registerDevelopmentSampleDataSources(self):
       checksums=['SHA256:c0743772587e2dd4c97d4e894f5486f7a9a202049c8575e032114c0a5c935c3b', 'SHA256:3243b62bde36b1db1cdbfe204785bd4bc1fbb772558d5f8cac964cda8385d470']
       )
 
+  def registerTestingDataSources(self):
+    """Fills in the sample data sources displayed only if developer mode is enabled."""
```
Comment (jcfr) [inline Modules/Scripted/SampleData/SampleData.py]: ``Fills in the sample data sources displayed only if developer mode is enabled.` -> `Register sample data sources used by SampleData self-test to test module functionalities.`

File: Modules/Scripted/SampleData/SampleData.py [hunk 10]
```diff
@@ -965,3 +983,13 @@ def test_categoryVisibility(self):
     self.assertFalse(widget.isCategoryVisible('BuiltIn'))
     widget.setCategoryVisible('BuiltIn', True)
     self.assertTrue(widget.isCategoryVisible('BuiltIn'))
+
+  def test_customDownloader(self):
+    logic = SampleDataLogic()
```
Comment (jcfr) [inline Modules/Scripted/SampleData/SampleData.py]: Remove `logic`

**Comment by jcfr** [comment]
I address the comments. There is one last issue, the exception raised can not be catched from the test because it happen a signal is triggered ... to workaroud this, we could add an ivar to the test object, associated the test object with CustomDownloader object and and ensure the test object is updated

**Comment by jcfr** [comment]
Closing. Integrated in [r28318](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=28318)

@lopezmt  Since I forgot to specify (--add-author-from) when committing, I added  a not the this wiki page to remind us to give you credit when transitioning the repository to GitHub.

**Pull request closed without merging**